### ISSUE TITLE
svg-tag for begin_src destroy the font lock in emacs 29

### ISSUE THREAD

**Issue by TimotheeMathieu** [rougier/svg-tag-mode#48]
Hello,

I recently updated to emacs 29.1 (GNU Emacs 29.1 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.38, cairo version 1.17.8)) and I noticed that svg-tag seemd to conflict with org font-lock in my config. 

Here is a minimal reproduction example:

`init.el`
```elisp
(setq straight-use-package-by-default t)

(defvar bootstrap-version)
(let ((bootstrap-file
       (expand-file-name "straight/repos/straight.el/bootstrap.el" user-emacs-directory))
      (bootstrap-version 6))
  (unless (file-exists-p bootstrap-file)
    (with-current-buffer
        (url-retrieve-synchronously
         "https://raw.githubusercontent.com/radian-software/straight.el/develop/install.el"
         'silent 'inhibit-cookies)
      (goto-char (point-max))
      (eval-print-last-sexp)))
  (load bootstrap-file nil 'nomessage))

(setq straight-use-package-by-default t)

(use-package svg-tag-mode)

(setq svg-tag-tags
      `(
         ;; source block from notebook-mode
 ("^#\\+begin_src\\( [a-zA-Z\-]+\\)" .  ((lambda (tag)
                                                  (svg-tag-make (upcase tag)
                                                                :face 'org-meta-line
                                                                :crop-left t))))
)
)

(add-hook 'text-mode-hook 'svg-tag-mode)
```

And then the file test.org
```
* Test
#+begin_src emacs-lisp
(message "Hello world")
#+end_src

#+RESULTS:
: Hello world
```

appears like this with svg-tag:

![image](https://github.com/rougier/svg-tag-mode/assets/30346931/6222071c-7d17-44ba-95e5-3d79f1a98ea3)

and like this if I have no svg-tag activated (which is equivalent to what I have with emacs -q)

![image](https://github.com/rougier/svg-tag-mode/assets/30346931/6c5b742e-04cf-4168-953b-db8ff0458b90)

This happens only with this src tag that worked well before the update to emacs 29, I also have a TODO svg-tag in my config and this one works well.

This may be linked with #47 as I have the same 
```
Error during redisplay: (jit-lock-function 1) signaled (wrong-type-argument stringp nil)
```
poping up in the messages buffer.

**Comment by yaqubroli** [comment]
Having a similar issue on mitsuharu's emacs-mac (railwaycat Homebrew tap, emacs 29), although it also resembles #27 (partially renders some formatting, leaves a large chunk of the file plain).

Here is the section of my config that sets up svg-tag-mode: https://github.com/yaqubroli/yaqubmacs/blob/master/rougier_tmp.org

![Screenshot 2023-08-19 at 2 48 06 pm](https://github.com/rougier/svg-tag-mode/assets/12129356/61009bbe-0e41-415c-a457-7930f6f30f47)

Initially I was going to say that it only renders partially when I evaluate the config section that installs svg-tag-mode, but *don't* restart emacs. When I restarted emacs before, it would show this file without any formatting whatsoever. However, in trying to replicate the issue, I haven't been able to get this specific file to display completely plainly. It does it with some files though (maybe shorter ones?)

I also get the `jit-lock-function` error like above and #47. Maybe, since the file is only rendering partially, there's some memory leak/allocation issue? It seems weird that it would start at the bottom of the file though.

**Comment by yaqubroli** [comment]
Ah, it appears that the file renders plainly when I enable ~jit-lock-debug-mode~. This is very weird; probably a bug in emacs itself?

![Screenshot 2023-08-19 at 3 21 45 pm](https://github.com/rougier/svg-tag-mode/assets/12129356/059de851-7f66-43fb-9d33-7df466f20470)

**Comment by rougier** [comment]
I think the jit-lock error comes from the added "display" property to font lock: [(add-to-list 'font-lock-extra-managed-props 'display)](https://github.com/rougier/svg-tag-mode/blob/main/svg-tag-mode.el#L285). 

If you comment it (you won't get tag obviously), do you still get errors?

**Comment by TimotheeMathieu** [comment]
I tried to comment out the "display" property that you linked to but I still have the same problem. 

Remark that, if that helps, using tags without regexp seems to not cause problem, i.e. using only this tag instead of the one from my first post:
```
 ("^#\\+begin_src" .  ((lambda (tag)
                                                  (svg-tag-make (upcase tag)
                                                                :face 'org-meta-line
                                                                :crop-left t))))
```
So the problem seems to come from regexp computation.

**Comment by rougier** [comment]
Can you try using the same regex without svg-tag (replace text with another string) ?

**Comment by TimotheeMathieu** [comment]
Yes, I tried and if I replace with string then the font lock works, there is no problem.

**Comment by rougier** [comment]
Could you try:
```
(add-to-list 'font-lock-extra-managed-props 'display)
(font-lock-add-keywords nil
  `(("\\(TODO\\):" 1 '(face nil display ,(svg-lib-tag "TODO")) prepend)))
```

**Comment by TimotheeMathieu** [comment]
From what I see it is meant to act on TODO, but with your code I don't see anything changing for TODO's.

I tried with and without `begin_src` svg-tags.

**Comment by TimotheeMathieu** [comment]
No I see no change. 
I am not sure what your code was meant to do but I guess you wanted to test that the TODO became an svg but it does not and there is nothing in the *Message* buffer.

**Comment by rougier** [comment]
Weird. Do you have SVG support? Can you try:

```emacs-lisp
;; TODO: test
(add-to-list 'font-lock-extra-managed-props 'display)
(font-lock-add-keywords nil
  `(("\\(TODO\\):" 1 '(face nil display "WORKS") prepend)))
```
If you edit the TODO, it shoudl display "WORKS". Maybe you'll need to setup a mode such as text-mode or lisp-mode (to have font-lock enabled)

**Comment by TimotheeMathieu** [comment]
I tried and indeed this replaces TODO with WORKS. It works, as you said it should.

**Comment by rougier** [comment]
This probably means the tag is nil is some situation. Can you try:
```
(setq svg-tag-tags
      `(
         ;; source block from notebook-mode
 ("^#\\+begin_src\\( [a-zA-Z\-]+\\)" .  ((lambda (tag)
     (message "Tag value: %s" tag)
                                                  (svg-tag-make "TEST"
                                                                :face 'org-meta-line
                                                                :crop-left t))))))```
```

**Comment by TimotheeMathieu** [comment]
I tried it: I got the message `Tag value: emacs-lisp` in the messsage buffer for my emacs-lisp source code but on the other hand, the text was without any font lock and there was not FAIL and no svg.

**Comment by rougier** [comment]
Can you try to using it with your original org file (the one that shows SVG errors)?

**Comment by TimotheeMathieu** [comment]
Yes. 
I added the code to the init.el and when I do `M-x svg-tag-mode` while in the org-mode file, the font-lock disappear from my test org-mode file and I get the following messages in the `*Messages*` buffer:
```
Svg-Tag mode enabled in current buffer
Tag value:  emacs-lisp
Error during redisplay: (jit-lock-function 1) signaled (wrong-type-argument stringp nil)
You can run the command ‘svg-tag-mode’ with M-x sv
Tag value:  emacs-lisp
```

**Comment by rougier** [comment]
Another try:
```emacs-lisp
#+begin_src emacs-lisp
(font-lock-add-keywords nil
         `(("lisp"
            0 `(face nil
                display (svg-lib-tag "LISP"))
        prepend)) 'append)
(font-lock-update)
#+end_src
```

**Comment by TimotheeMathieu** [comment]
This works, transform the lisp  to LISP when executed and message is `(jit-lock-bounds 1 . some-number)` with the some-number being an integer different every time I execute.

**Comment by TimotheeMathieu** [comment]
I tried to change the library with some intuition of what things should do, but I am really not fluent enough in elisp and in particular in font lock to be able to do a PR.
So thanks for asking, but I don't know how to solve it, so I can't do a PR. 
I am available to test any PR that comes up though.

**Comment by rougier** [comment]
No problem. I'll try to make the PR. I leave the issue open until it is fixed. Don't hesitate to remind me.
